### ISSUE TITLE
X11 and static Qt

### ISSUE THREAD

**Issue by darealshinji** [mbunkus/mkvtoolnix#1898]
Can you make it a bit easier to link against static Qt on X11/Linux systems? Right now the xcb plugin won't be added. Maybe an environment variable or configure option to set the platform plugin libraries manually would help.

These are the changes I used for my build:
``` diff
--- a/src/info/qt_ui.cpp
+++ b/src/info/qt_ui.cpp
@@ -42,6 +42,10 @@
 #  include <QtPlugin>
 Q_IMPORT_PLUGIN(QWindowsIntegrationPlugin);
 
+# else
+#  include <QtPlugin>
+Q_IMPORT_PLUGIN(QXcbIntegrationPlugin);
+
 # endif  // SYS_APPLE, SYS_WINDOWS
 #endif   // HAVE_STATIC_QT
 
--- a/src/mkvtoolnix-gui/mkvtoolnix_gui.cpp
+++ b/src/mkvtoolnix-gui/mkvtoolnix_gui.cpp
@@ -21,6 +21,10 @@
 #  include <QtPlugin>
 Q_IMPORT_PLUGIN(QWindowsIntegrationPlugin);
 
+# else
+#  include <QtPlugin>
+Q_IMPORT_PLUGIN(QXcbIntegrationPlugin);
+
 # endif  // SYS_APPLE, SYS_WINDOWS
 #endif   // HAVE_STATIC_QT
 using namespace mtx::gui;
--- a/ac/qt5.m4
+++ b/ac/qt5.m4
@@ -204,7 +204,7 @@
         *mingw*)  wanted_plugin=windows ;;
         *apple*)  wanted_plugin=cocoa   ;;
         *darwin*) wanted_plugin=cocoa   ;;
-        *)        wanted_plugin=        ;;
+        *)        wanted_plugin=xcb     ;;
       esac
 
       if test x"$wanted_plugin" != x; then
@@ -220,7 +220,12 @@
           AC_MSG_RESULT(no: the platform plugins directory could not be found)
           have_qt=no
         else
-          QT_LIBS="-L$plugins_dir/platforms -lq$wanted_plugin $QT_LIBS"
+          if test x"$wanted_plugin" = x"xcb"; then
+            QT_LIBS="$(echo $(grep '^QMAKE_PRL_LIBS' "$dir/../plugins/platforms/libqxcb.prl" | sed -e 's/\t/\n/g; s/ /\n/g' | grep '^-l')) $QT_LIBS"
+            QT_LIBS="-L$plugins_dir/platforms -lqxcb $QT_LIBS"
+          else
+            QT_LIBS="-L$plugins_dir/platforms -lq$wanted_plugin $QT_LIBS"
+          fi
         fi
       fi
 
```

**Comment by mbunkus** [comment]
I don't want to spend time on implementing this myself, to be honest. I will apply patches, though.

The current patch isn't acceptable to me as always importing the Xcb seems… at least unnecessary (as in I don't need to import it at the moment in a normal shared build) if not flat-out wrong. I'd be fine with a configure option that forces the plugin to be imported, though.

**Comment by darealshinji** [comment]
The first two patches should be ok, as the plugin is only imported if `HAVE_STATIC_QT` was defined. Maybe an extra check could be done to see if we're using X11. The qt5.m4 patch is a dirty hack. I'll see if I can add an option to manually override the platform plugin libraries if the user needs to do so.

**Comment by mbunkus** [comment]
I'd really prefer the C++ part to be gated by an `#if defined(…)`, with the symbol to check before being dependent on the `configure` option. Ideally:

* `configure` checks which plugin to use and sets e.g. `MTX_QT_PLATFORM_PLUGIN` to the name of the plugin to use.
* The C++ part adds checks such as `#if defined(MTX_QT_PLATFORM_PLUIN) && (MTX_QT_PLATFORM_PLUIN == "xcb")` instead of the current OS-dependent variants.
* There are currently two places where the plugins have to be imported (`info/qt_ui.cpp` and `mkvtoolnix-gui/mkvtoolnix_gui.cpp`). It'd be great if your changes could move that code to a header file such as `common/qt_plugins.h`, and `info/qt_ui.cpp` and `mkvtoolnix-gui/mkvtoolnix_gui.cpp` should only include that header.

Alternatively define something like `MTX_QT_PLATFORM_PLUGIN_XCB`, `MTX_QT_PLATFORM_PLUGIN_COCOA` etc. and use `#if deined(MTX_QT_PLATFORM_PLUGIN_XCB)` (for the other two, too, please).

**Comment by mbunkus** [comment]
I've just pushed a commit that refactors how the static plugins are detected and loaded. The commit also adds the inclusion of the XCB platform integration plugin on platforms other than Windows and macOS. Please give it a try.

**Comment by mbunkus** [comment]
Ping @darealshinji . Have you had time to give the current sources a try?

**Comment by darealshinji** [comment]
Configure fails to link the conftest binary, so Qt is not enabled:
 https://gist.github.com/darealshinji/8a89317bf488981e92f4aedeee396d07#file-config-log-L1055

**Comment by mbunkus** [comment]
Too bad, but thanks for the feedback. As I've said above I don't want to spend time on this. Therefore I'm going to close it as WONTFIX. The current code does work with static Qt for Windows and macOS, and that suffices for me.

If you can come up with a patch that fixes the issue, then I'll gladly apply it.
